include/ofi_mr.h

```diff
@@ -141,4 +145,46 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
 		      size_t len, uint64_t key, uint64_t access,
 		      void **context);
 
+
+/*
+ * Memory registration cache
+ */
+
+struct ofi_mr_entry {
+	struct iovec		iov;
+	uint64_t		access;	/* TODO */
+	unsigned int		retired:1;
+	int			use_cnt;
+	struct dlist_entry	lru_entry;
+	uint8_t			data[];
+};
+
+struct ofi_mr_cache {
+	struct util_domain	*domain;
+	size_t			size;
+	size_t			entry_data_size;
+
+	RbtHandle		mr_tree;
+	struct dlist_entry	lru_list;
+
+	uint64_t		cached_cnt;
+	uint64_t		search_cnt;
+	uint64_t		delete_cnt;
+	uint64_t		hit_cnt;
+
+	int			(*add_region)(struct ofi_mr_cache *cache,
+					       struct ofi_mr_entry *entry);
+	void			(*delete_region)(struct ofi_mr_cache *cache,
+						 struct ofi_mr_entry *entry);
+};
```

Comment:
The flush call was originally provided to allow the underlying implementation to remove stale entries in the event that the NIC was resource bound in terms of memory registrations. Specifically with the GNI provider, there is a hard limit of 4096 memory registrations. As such, if stale entries needed to be discarded to free space, the provider could call flush to release stale registrations and attempt to register again. I believe such a mechanism is necessary for more providers than just the GNI provider. Flushing generally means going through the LRU and removing a stale entry to free space for the new registration. Flushing would only be necessary if a current or stale entry couldn't satisfy the registration request.

Reply:
Exposing flush should be straightforward.  Are we needing to flush all lru entries, or just one?

Reply:
I always included a number argument and left it to the upper layer to decide. 1 to x, or -1 to flush them all.

Reply:
This is an internal interface, so we can change it anytime, but really only need to be concerned with what the callers will actually use.  Flush-one seems like it would be the best option unless we end up adding some sort of time stamps.  (I'd like to avoid the overhead of taking time stamps.)